### ISSUE TITLE
add --list-target-files option

### ISSUE THREAD

**Issue by flyerhzm** [eslint/eslint#11005]
**The version of ESLint you are using.**

5.7.0

**The problem you want to solve.**

I'm working on https://awesomecode.io, our users may not use `eslint` directly on their shell, we have a requirement to list target files so that our users can add ignore patterns properly.

**Your take on the correct solution to problem.**

We can add prefix `DEBUG=eslint:linter` to get target files, but it's very slow and its output contains too many verbose text.

My solution is to add `--list-target-files` option, then it gets all target files without parsing files, so it's super fast. Its output is straightforward, only lists the target filenames.

**Comment by not-an-aardvark** [comment]
Hi, thanks for creating an issue.

If I'm understanding correctly, the problem you're trying to solve is to figure out how ESLint resolves your globs and lints your files, for debugging purposes. It's already somewhat possible to do this with the `DEBUG` environment variable. For example, if I run something like this:

`DEBUG=eslint:linter eslint tools/`

Then ESLint generates output like this:

<details>
<summary>Output</summary>

```
  eslint:linter Linting code for /path/to/tools/eslint-fuzzer.js (pass 1) +0ms
  eslint:linter Generating fixed text for /path/to/tools/eslint-fuzzer.js (pass 1) +189ms
  eslint:linter Linting code for /path/to/tools/fuzzer-runner.js (pass 1) +1ms
  eslint:linter Generating fixed text for /path/to/tools/fuzzer-runner.js (pass 1) +30ms
  eslint:linter Linting code for /path/to/tools/internal-rules/consistent-docs-description.js (pass 1) +1ms
  eslint:linter Generating fixed text for /path/to/tools/internal-rules/consistent-docs-description.js (pass 1) +66ms
  eslint:linter Linting code for /path/to/tools/internal-rules/consistent-docs-url.js (pass 1) +1ms
  eslint:linter Generating fixed text for /path/to/tools/internal-rules/consistent-docs-url.js (pass 1) +36ms
  eslint:linter Linting code for /path/to/tools/internal-rules/consistent-meta-messages.js (pass 1) +0ms
  eslint:linter Generating fixed text for /path/to/tools/internal-rules/consistent-meta-messages.js (pass 1) +29ms
  eslint:linter Linting code for /path/to/tools/internal-rules/multiline-comment-style.js (pass 1) +0ms
  eslint:linter Generating fixed text for /path/to/tools/internal-rules/multiline-comment-style.js (pass 1) +16ms
  eslint:linter Linting code for /path/to/tools/internal-rules/no-invalid-meta.js (pass 1) +1ms
  eslint:linter Generating fixed text for /path/to/tools/internal-rules/no-invalid-meta.js (pass 1) +50ms
  eslint:linter Linting code for /path/to/tools/internal-rules/no-useless-catch.js (pass 1) +1ms
  eslint:linter Generating fixed text for /path/to/tools/internal-rules/no-useless-catch.js (pass 1) +21ms
  eslint:linter Linting code for /path/to/tools/internal-testers/event-generator-tester.js (pass 1) +0ms
  eslint:linter Generating fixed text for /path/to/tools/internal-testers/event-generator-tester.js (pass 1) +17ms
  eslint:linter Linting code for /path/to/tools/internal-testers/test-parser.js (pass 1) +0ms
  eslint:linter Generating fixed text for /path/to/tools/internal-testers/test-parser.js (pass 1) +31ms
  eslint:linter Linting code for /path/to/tools/update-unicode-utils.js (pass 1) +1ms
  eslint:linter Generating fixed text for /path/to/tools/update-unicode-utils.js (pass 1) +21ms
```

</details>

In other words, I'm not sure adding a new command-line flag is necessary given that there's already a way to do this. Does the ability above satisfy your use case?

**Comment by nzakas** [comment]
Agree with @not-an-aardvark, this is precisely where the debug output is useful.

**Comment by flyerhzm** [comment]
@not-an-aardvark glad to know eslint already can handle this, but it's not good enough.

1. it's too slow to use DEBUG way, here's an example to run on eslint itself.

DEBUG solution

```
DEBUG=eslint:linter ./bin/eslint.js .  28.36s user 1.09s system 141% cpu 20.883 total
```

--list-target-files solution

```
./bin/eslint.js --list-target-files .  0.43s user 0.14s system 110% cpu 0.516 total
```

2. it outputs too many texts, ('Linting code for', 'Generating fixed text for', warnings and error, etc.), --list-target-files solution only outputs files and it's easy for other tool to analyze STDOUT

**Comment by not-an-aardvark** [comment]
From your initial description, I got the impression that this is mainly intended for debugging, in which case the performance would probably not be too important, and it wouldn't be necessary to analyze the output with tools. Could you clarify your use case for having the output of this command parseable by tools?

**Comment by flyerhzm** [comment]
@not-an-aardvark okay, I'm working on https://awesomecode.io, we use eslint internally and we want to let users know what files will be parsed and what are not, it's much easier to parse the output with `--list-target-files` option.

**Comment by nzakas** [comment]
@flyerhzm in that case, can you please go back and edit your initial comment? One sentence isn't really enough for us to determine if we'd like to move forward with this suggestion. Please describe the complete problem you are trying to solve, including details about how you use ESLint in your project currently and how you might implement the solution you are proposing. Thanks!

**Comment by flyerhzm** [comment]
@nzakas updated

**Comment by flyerhzm** [comment]
@not-an-aardvark yes, it can be a solution for us, I will use the API if this PR is not merged.

**Comment by platinumazure** [comment]
@not-an-aardvark Do you know if `CLIEngine#resolveFileGlobPatterns` will take ignore patterns into account? My assumption is it won't. We do have `CLIEngine#isPathIgnored`, so maybe the solution is to resolve the glob patterns, call the glob utility to get a list of paths, and then filter that list to remove ignored paths.

**Comment by kaicataldo** [comment]
From looking at the code, it doesn't look like it takes ignored paths into account. This [method](https://github.com/eslint/eslint/blob/6009239042cb651bc7ca6b8c81bbe44c40327430/lib/util/glob-utils.js#L187) (which we don't currently expose publicly) looks to be what this issue is asking for. I'm not convinced we need a new command-line flag either, but I wouldn't be opposed to exposing this in the Node.js API somehow.
